sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs

```diff
@@ -0,0 +1,130 @@
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotMarkEnumsWithFlags : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4070";
+        private const string MessageFormat = "Remove the 'FlagsAttribute' from this enum.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var enumDeclaration = (EnumDeclarationSyntax)c.Node;
+                    var enumSymbol = c.SemanticModel.GetDeclaredSymbol(enumDeclaration);
+
+                    if (!enumDeclaration.HasFlagsAttribute(c.SemanticModel) ||
+                        enumDeclaration.Identifier.IsMissing ||
+                        enumSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var membersWithValues = enumSymbol.GetMembers()
+                        .OfType<IFieldSymbol>()
+                        .Select(member => new { Member = member, Value = GetEnumValueOrDefault(member) })
+                        .OrderByDescending(tuple => tuple.Value)
+                        .ToList();
+
+                    var allValues = membersWithValues.Select(x => x.Value)
+                        .OfType<ulong>()
+                        .ToList();
+
+                    var invalidMembers = membersWithValues.Where(tuple => IsInvalidFlagValue(tuple.Value, allValues))
+                        .Select(tuple => tuple.Member.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax().GetLocation())
+                        .WhereNotNull()
+                        .ToList();
+
+                    if (invalidMembers.Count > 0)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, enumDeclaration.Identifier.GetLocation(),
+                            additionalLocations: invalidMembers));
+                    }
+                }, SyntaxKind.EnumDeclaration);
+        }
+
+        private static ulong? GetEnumValueOrDefault(IFieldSymbol enumMember)
+        {
+            ulong longValue;
+            if (!enumMember.HasConstantValue ||
+                !ulong.TryParse(enumMember.ConstantValue.ToString(), out longValue))
+            {
+                return null;
```

Comment:
This is the kind of case when I'm tempted to add Debug.Assert. WDYT? Could enum ever not have a value?